### PR TITLE
AAC: consensus tests for `EarlyReturnError` variants

### PR DESCRIPTION
**Pull request by jacinta-stacks**
Closes https://github.com/stacks-network/stacks-core/issues/6702

### CODE REVIEW

File: stackslib/src/chainstate/tests/early_return_tests.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    );
+}
+
+/// Error: [`EarlyReturnError::UnwrapFailed`]
+/// Caused by: attempting to `try!` unwrap an `None` optional at call time.
+/// Outcome: block accepted
+/// Note: [`clarity::vm::callables::DefinedFunction::execute_apply`] converts [`EarlyReturnError::UnwrapFailed`]
+/// into a successful return wrapping the internal thrown value.
+#[test]
+fn native_try_ret_none_ccall() {
```
Comment (federico-stacks) [inline stackslib/src/chainstate/tests/early_return_tests.rs:123]: Something I noticed is that this test case is the only one (among the contract-call ones) where the transaction result doesn't report if the tx has been committed or not.

Here an example extracted from `blockstack_lib__chainstate__tests__early_return_tests__native_try_ret_none_ccall` where the `committed` field seems missing.

```
      ExpectedTransactionOutput(
        tx: "ContractCall(address: ST1AW6EKPGT61SQ9FNVDS17RKNWT8ZP582VF9HSCP, contract_name: unwrap-try-opt-Epoch3_3-Clarity4, function_name: trigger, function_args: [[Optional(OptionalData { data: None })]])",
        vm_error: "None [NON-CONSENSUS BREAKING]",
        return_type: Optional(OptionalData(
          data: None,
        )),
```

Comment (francesco-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: good catch, that's definitely an oddity

**Comment by francesco-stacks** [comment]
> Just a quick note: this isn’t related to the PR itself, but something I noticed while reviewing it.
> 
> If I remember correctly, we had agreed to rename the variant ClarityError::Interpreter(VmExecutionError) to ClarityError::Execution(VmExecutionError). However, it’s still present as ClarityError::Interpreter.
> 
> Did we change our minds on that at some point (and I’m just forgetting), or did the rename simply never happen (or lost in one of many merge conflicts)?
> 
> @jacinta-stacks @francesco-stacks

The name didn't ring a bell, but I went through old conversations and files, and you are correct. It was once proposed to be renamed that way (by me D: ). I am not sure If I missed that during the renaming or there was another reason behind. Anyway, now that there are no other conflicting variants or enums with the same name, `Interpreter` doesnt' bother me very much, but I'd be supportive if you feel the name would be better

**Comment by jacinta-stacks** [comment]
> Just a quick note: this isn’t related to the PR itself, but something I noticed while reviewing it.
> 
> If I remember correctly, we had agreed to rename the variant ClarityError::Interpreter(VmExecutionError) to ClarityError::Execution(VmExecutionError). However, it’s still present as ClarityError::Interpreter.
> 
> Did we change our minds on that at some point (and I’m just forgetting), or did the rename simply never happen (or lost in one of many merge conflicts)?
> 
> @jacinta-stacks @francesco-stacks

Yeah i thought for some reason it had been changed already but apparently not. I think it got lost in a merge as part of the VmExecution rename. But I am good either way.

File: stackslib/src/chainstate/tests/early_return_tests.rs (hunk 1 above)
Comment (jacinta-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Right so this is because it returns the wrapped value which in this case is an optional type. i can simluate this exact same result in `native_unwrap_or_ret_none_ccall` as well if I change it to this.
```rust
#[test]
fn native_unwrap_or_ret_none_ccall() {
    contract_call_consensus_test!(
        contract_name: "unwrap-opt",
        contract_code: "
            (define-read-only (trigger (opt (optional bool)))
                (begin
                    (unwrap! opt (some false))
                    (some true)
                )
            )
        ",
        function_name: "trigger",
        function_args: &[ClarityValue::none()],
    );
}
```

Comment (jacinta-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Basically I would need try! to throw a resp but I am specifically wanting it to throw  by trying on an optional. So I mean its only odd because all other tests throw a response whereas this test directly throws an optional if that makes sense.

Comment (federico-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: If it this what you wanted to achive, ok. For sure it is a valid clarity `Value`. 

One more question: In this case, how we can understand if the tx has failed?

Comment (francesco-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: What I find strange is that this is the first ever contract call that doesn't include the `committed: true/false`, not a  problem if it's not there, but I was wondering if this was a bug in the implementation of this particular code path in clarity

Comment (jacinta-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: A good question. We know the block is accepted but nothing about the transaction succeeding or not. @brice-stacks how do you know if a transaction succeeds or not if the transaction does not return a response?

Comment (brice-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: I don't think it's possible to include a transaction in a block that calls anything but a public function, which must return a `response` type. Let me verify that though to be sure...

Comment (brice-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Right, you will get a `MemPoolRejection::NoSuchPublicFunction` if you attempt to submit a transaction which calls a readonly function.

Comment (jacinta-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Ah this makes a lot of sense. Let me see if there are tests in here that needlessly call read only functions (some may be required to trigger the intended code paths)

Comment (jacinta-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Ah interesting though is that if we simulate a block that actually includes these...they do seem to get mined without issue. i.e. if someone bypasses somehow mempool rejection, it does not invalidate a block.  @aaronb-stacks is this acceptable?

Comment (aaronb-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Yeah -- that's okay -- it's forced read-only, so this is okay, but it is a weird corner case that I think we'll want to keep a test around for.

Comment (brice-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: A read-only function could only fail if it triggered a runtime error, which I think would cause it to have a response result.

Comment (brice-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Yeah, I just tested with an arithmetic underflow and the return value is an `(err none)`.

Comment (jacinta-stacks) [reply stackslib/src/chainstate/tests/early_return_tests.rs:123]: Added a sep test in consensus.rs to test what happens when a read only call is mined. I also changed the other tests to not directly call a read only so we know for sure if the transaction was committed or not. 

See [147105c](https://github.com/stacks-network/stacks-core/pull/6736/commits/147105ce118e1c5fa4683a03b3536e33ada192ff)
